File: Viewer/py/lib/graph.py
"""Python port of ts/lib/graph.ts — the module/edge/vertex connectivity graph.

Everything outside the app addresses LEDs in module/edge terms. This graph maps
between tiles, edges, ends and vertices so higher-level (hex) authoring can be
translated down to module edges.
"""
from __future__ import annotations

from dataclasses import dataclass, replace
from enum import Enum
from typing import ClassVar, Dict, Iterator, Tuple


class EdgeClass(str, Enum):
    A = "A"
    B = "B"
    C = "C"
    D = "D"
    E = "E"
    F = "F"


class VertexClass(str, Enum):
    ABF = "ABF"
    ABC = "ABC"
    CDE = "CDE"
    DEF = "DEF"


@dataclass(frozen=True)
class TileRef:
    x: int
    y: int

    # ClassVar so they aren't __init__ fields; the singletons are attached below.
    THIS: ClassVar["TileRef"]
    LEFT: ClassVar["TileRef"]
    RIGHT: ClassVar["TileRef"]
    UP: ClassVar["TileRef"]
    DOWN: ClassVar["TileRef"]

    @staticmethod
    def from_xy(x: int, y: int) -> "TileRef":
        return TileRef(x, y)

    def offset(self, a):
        """Translate any {tile: TileRef} ref by this tile's offset."""
        return replace(a, tile=a.tile.add(self))

    def negate(self) -> "TileRef":
        return TileRef(-self.x, -self.y)

    def add(self, tr: "TileRef") -> "TileRef":
        return TileRef(self.x + tr.x, self.y + tr.y)

    def edge(self, c: EdgeClass) -> "EdgeRef":
        return EdgeRef(c, self)

    def vertex(self, v: VertexClass) -> "VertexRef":
        return VertexRef(v, self)

    def top_end(self, c: EdgeClass) -> "EndRef":
        return EndRef(c, self, True)

    def bottom_end(self, c: EdgeClass) -> "EndRef":
        return EndRef(c, self, False)


# dataclass turns the annotations above into fields; strip the constant markers
# and attach the real singletons.
for _name in ("THIS", "LEFT", "RIGHT", "UP", "DOWN"):
    TileRef.__dataclass_fields__.pop(_name, None)
TileRef.THIS = TileRef(0, 0)
TileRef.LEFT = TileRef(-1, 0)
TileRef.RIGHT = TileRef(1, 0)
TileRef.UP = TileRef(0, 1)
TileRef.DOWN = TileRef(0, -1)


@dataclass(frozen=True)
class VertexRef:
    vertex_class: VertexClass
    tile: TileRef

    @staticmethod
    def ends_cw(v: "VertexRef") -> Tuple["EndRef", "EndRef", "EndRef"]:
        """Clockwise ends around the vertex; index 0 is always the vertical."""
        s = VERTEX_END_SETS[v.vertex_class]
        points_up = not s.v.top
        order = (s.v, s.r, s.l) if points_up else (s.v, s.l, s.r)
        return tuple(v.tile.offset(x) for x in order)  # type: ignore[return-value]


@dataclass(frozen=True)
class EdgeRef:
    edge_class: EdgeClass
    tile: TileRef

    @staticmethod
    def ends(e: "EdgeRef") -> Tuple["EndRef", "EndRef"]:
        return (EndRef(e.edge_class, e.tile, True),
                EndRef(e.edge_class, e.tile, False))

    @staticmethod
    def vertex_pair(e: "EdgeRef") -> "EdgeVertexPair":
        p = EDGE_VERTEX_PAIRS[e.edge_class]
        return EdgeVertexPair(e.tile.offset(p.top), e.tile.offset(p.bottom))


@dataclass(frozen=True)
class EndRef(EdgeRef):
    top: bool

    @staticmethod
    def other(er: "EndRef") -> "EndRef":
        return replace(er, top=not er.top)

    @staticmethod
    def vertex(e: "EndRef") -> VertexRef:
        vp = EdgeRef.vertex_pair(e)
        return vp.top if e.top else vp.bottom

    @staticmethod
    def lr(er: "EndRef") -> Tuple["EndRef", "EndRef"]:
        cw = VertexRef.ends_cw(EndRef.vertex(er))
        if cw[0].edge_class == er.edge_class:
            return (cw[1], cw[2])
        elif cw[1].edge_class == er.edge_class:
            return (cw[2], cw[0])
        else:
            return (cw[0], cw[1])


@dataclass(frozen=True)
class EndSet:
    v: EndRef
    l: EndRef
    r: EndRef


@dataclass(frozen=True)
class EdgeVertexPair:
    top: VertexRef
    bottom: VertexRef


VERTEX_END_SETS: Dict[VertexClass, EndSet] = {
    VertexClass.ABC: EndSet(
        v=TileRef.THIS.top_end(EdgeClass.C),
        l=TileRef.LEFT.bottom_end(EdgeClass.B),
        r=TileRef.THIS.bottom_end(EdgeClass.A),
    ),
    VertexClass.ABF: EndSet(
        v=TileRef.UP.bottom_end(EdgeClass.F),
        l=TileRef.THIS.top_end(EdgeClass.A),
        r=TileRef.THIS.top_end(EdgeClass.B),
    ),
    VertexClass.CDE: EndSet(
        v=TileRef.THIS.bottom_end(EdgeClass.C),
        l=TileRef.LEFT.top_end(EdgeClass.E),
        r=TileRef.THIS.top_end(EdgeClass.D),
    ),
    VertexClass.DEF: EndSet(
        v=TileRef.THIS.top_end(EdgeClass.F),
        l=TileRef.THIS.bottom_end(EdgeClass.D),
        r=TileRef.THIS.bottom_end(EdgeClass.E),
    ),
}

EDGE_VERTEX_PAIRS: Dict[EdgeClass, EdgeVertexPair] = {
    EdgeClass.A: EdgeVertexPair(VertexRef(VertexClass.ABF, TileRef.THIS),
                                VertexRef(VertexClass.ABC, TileRef.THIS)),
    EdgeClass.B: EdgeVertexPair(VertexRef(VertexClass.ABF, TileRef.THIS),
                                VertexRef(VertexClass.ABC, TileRef.RIGHT)),
    EdgeClass.C: EdgeVertexPair(VertexRef(VertexClass.ABC, TileRef.THIS),
                                VertexRef(VertexClass.CDE, TileRef.THIS)),
    EdgeClass.D: EdgeVertexPair(VertexRef(VertexClass.CDE, TileRef.THIS),
                                VertexRef(VertexClass.DEF, TileRef.THIS)),
    EdgeClass.E: EdgeVertexPair(VertexRef(VertexClass.CDE, TileRef.RIGHT),
                                VertexRef(VertexClass.DEF, TileRef.THIS)),
    EdgeClass.F: EdgeVertexPair(VertexRef(VertexClass.DEF, TileRef.THIS),
                                VertexRef(VertexClass.ABF, TileRef.DOWN)),
}


@dataclass(frozen=True)
class HexGridCoord:
    x: int
    y: int

    @staticmethod
    def ends(gc: "HexGridCoord") -> Iterator[EndRef]:
        """The 12 ends (top/bottom of each of 6 edges) around a hexagon."""
        if gc.y % 2 == 0:
            vertex = TileRef(gc.x, gc.y // 2).vertex(VertexClass.DEF)
        else:
            vertex = TileRef(gc.x + 1, gc.y // 2).vertex(VertexClass.ABC)

        below = VertexRef.ends_cw(vertex)[0]     # top of the vertical below the vertex
        current = EndRef.lr(below)[1]

        for _ in range(6):
            yield current
            current = EndRef.other(current)
            yield current
            current = EndRef.lr(current)[0]

File: Viewer/ts/lib/graph.ts
export enum EdgeClass {
    A = "A",
    B = "B",
    C = "C",
    D = "D",
    E = "E",
    F = "F"
}

export enum VertexClass {
    ABF = "ABF",
    ABC = "ABC",
    CDE = "CDE",
    DEF = "DEF"
}

export class TileRef {

    constructor(
        public x: number,
        public y: number) { }

    static readonly THIS = new TileRef(0, 0);
    static readonly LEFT = new TileRef(-1, 0);
    static readonly RIGHT = new TileRef(1, 0);
    static readonly UP = new TileRef(0, 1);
    static readonly DOWN = new TileRef(0, -1);

    static readonly WIDTH = 1;
    static readonly HEIGHT = 1.732;

    static readonly RADIUS = 0.577;

    /* Location of main hex center within a tile: */

    static readonly ORIGIN_OFFSET_X = 0.5;
    static readonly ORIGIN_OFFSET_Y = 2 * TileRef.RADIUS;


    static from(x: number, y: number) {
        return new TileRef(
            x,y
        )
    }

    static from_physical(x: number, y: number) {

        return TileRef.from(
            Math.floor((x - TileRef.ORIGIN_OFFSET_X) / TileRef.WIDTH),
            Math.floor((y - TileRef.ORIGIN_OFFSET_Y) / TileRef.HEIGHT)
        )
    }

    offset<const T extends { tile: TileRef }>(a: T): T {
        const b = { ...a };
        b.tile = b.tile.add(this);
        return b;
    }

    negate(): TileRef {
        return new TileRef(
            -this.x,
            -this.y
        )
    }

    add(tr: TileRef): TileRef {
        return new TileRef(
            this.x + tr.x,
            this.y + tr.y
        )
    }

    edge(c: EdgeClass): EdgeRef {
        return {
            edge_class: c,
            tile: this
        }
    }

    vertex(v: VertexClass): VertexRef {
        return {
            tile: this,
            vertex_class: v
        };
    }

    top_end(c: EdgeClass): EndRef {
        return {
            edge_class: c, tile: this,
            top: true
        }
    }

    bottom_end(c: EdgeClass): EndRef {
        return {
            edge_class: c, tile: this,
            top: false
        }
    }

    physical(): [number, number] {
        // Origin of a tile is at the main hex center point, which is at 0.5, 0.577
        // Width of a tile is '1'
        // Height of a tile is 1.732
        return [this.x * TileRef.WIDTH + TileRef.ORIGIN_OFFSET_X,
                this.y * TileRef.HEIGHT + TileRef.ORIGIN_OFFSET_Y]

    }
}
export type VertexRef = {
    vertex_class: VertexClass
    tile: TileRef
}

export type EdgeRef = {
    edge_class: EdgeClass
    tile: TileRef
}

export type EndRef = {
    top: boolean,
} & EdgeRef



// A vertex class has a number of endrefs
export type EndSet = {
    v: EndRef,
    l: EndRef,
    r: EndRef,
}


const VERTEX_END_SETS: { [key in VertexClass]: EndSet } = {
    [VertexClass.ABC]: {
        v: TileRef.THIS.top_end(EdgeClass.C),
        l: TileRef.LEFT.bottom_end(EdgeClass.B),
        r: TileRef.THIS.bottom_end(EdgeClass.A)
    }, [VertexClass.ABF]: {
        v: TileRef.UP.bottom_end(EdgeClass.F),
        l: TileRef.THIS.top_end(EdgeClass.A),
        r: TileRef.THIS.top_end(EdgeClass.B)
    }, [VertexClass.CDE]: {
        v: TileRef.THIS.bottom_end(EdgeClass.C),
        l: TileRef.LEFT.top_end(EdgeClass.E),
        r: TileRef.THIS.top_end(EdgeClass.D)
    }, [VertexClass.DEF]: {
        v: TileRef.THIS.top_end(EdgeClass.F),
        l: TileRef.THIS.bottom_end(EdgeClass.D),
        r: TileRef.THIS.bottom_end(EdgeClass.E)
    }
}


export const VertexRef = {
    // clockwise ends. 0 is always the vertical.
    ends_cw(v: VertexRef): [EndRef, EndRef, EndRef] {
        const set = VERTEX_END_SETS[v.vertex_class];

        const points_up = !set.v.top;

        return (points_up ? [
            set.v,
            set.r,
            set.l
        ] : [
            set.v,
            set.l,
            set.r
        ]).map(x => v.tile.offset(x)) as [EndRef, EndRef, EndRef];
    },

    // Physical location
    physical(v: VertexRef): [number, number] {

        // Center of main hex
        const center_pos = v.tile.physical();

        // Offsets from center
        const offs = TILE_VERTEX_OFFSETS[v.vertex_class];

        return [
            center_pos[0] + offs[0],
            center_pos[1] + offs[1]
        ]
    }
}

const TILE_VERTEX_OFFSETS: { [key in VertexClass]: [number, number] } = {
    [VertexClass.ABC]: [-TileRef.RADIUS, TileRef.RADIUS/2],
    [VertexClass.ABF]: [0, TileRef.RADIUS],
    [VertexClass.CDE]: [-TileRef.RADIUS, -TileRef.RADIUS/2],
    [VertexClass.DEF]: [0, -TileRef.RADIUS]
}



type EdgeVertexPair = {
    top: VertexRef,
    bottom: VertexRef
}

const EDGE_VERTEX_PAIRS: { [index in EdgeClass]: EdgeVertexPair } = {
    [EdgeClass.A]: {
        top: { vertex_class: VertexClass.ABF, tile: TileRef.THIS },
        bottom: { vertex_class: VertexClass.ABC, tile: TileRef.THIS }
    }, [EdgeClass.B]: {
        top: { vertex_class: VertexClass.ABF, tile: TileRef.THIS },
        bottom: { vertex_class: VertexClass.ABC, tile: TileRef.RIGHT }
    }, [EdgeClass.C]: {
        top: { vertex_class: VertexClass.ABC, tile: TileRef.THIS },
        bottom: { vertex_class: VertexClass.CDE, tile: TileRef.THIS }
    }, [EdgeClass.D]: {
        top: { vertex_class: VertexClass.CDE, tile: TileRef.THIS },
        bottom: { vertex_class: VertexClass.DEF, tile: TileRef.THIS }
    }, [EdgeClass.E]: {
        top: { vertex_class: VertexClass.CDE, tile: TileRef.RIGHT },
        bottom: { vertex_class: VertexClass.DEF, tile: TileRef.THIS }
    }, [EdgeClass.F]: {
        top: { vertex_class: VertexClass.DEF, tile: TileRef.THIS },
        bottom: { vertex_class: VertexClass.ABF, tile: TileRef.DOWN }
    }
}

export const EdgeRef = {
    ends(e: EdgeRef): [EndRef, EndRef] {
        return [
            { ...e, top: true },
            { ...e, top: false }
        ]
    },
    vertex_pair(e: EdgeRef): EdgeVertexPair {
        const edge_vertexes = EDGE_VERTEX_PAIRS[e.edge_class];
        return {
            top: e.tile.offset(edge_vertexes.top),
            bottom: e.tile.offset(edge_vertexes.bottom)
        }
    },
    top_end(er: EdgeRef) {
        return {...er, top: true}
    },
    bottom_end(er: EdgeRef) {
        return {...er, top: false}
    }
}

export const EndRef = {
    other(er: EndRef): EndRef {
        return { ...er, top: !er.top }
    },
    vertex(e: EndRef): VertexRef {
        // endref just says the edge, and whether it's at the top or bottom

        // Get top and bottom vertexes for this edge
        const edge_vertexes = EdgeRef.vertex_pair(e);

        // Return the one we're interested in
        return e.top ? edge_vertexes.top : edge_vertexes.bottom;
    },
    lr(er: EndRef): [EndRef, EndRef] {

        const vertex = EndRef.vertex(er);

        const cw = VertexRef.ends_cw(vertex);

        if (cw[0].edge_class == er.edge_class) {
            return [cw[1], cw[2]];
        } else if (cw[1].edge_class == er.edge_class) {
            return [cw[2], cw[0]];
        } else {
            return [cw[0], cw[1]];
        }
    },

    // Vector pointing from this end to the opposite
    physical_to_next(er: EndRef): [number, number] {
        const from = VertexRef.physical(this.vertex(er));
        const to = VertexRef.physical(this.vertex(this.other(er)));

        return [
            to[0] - from[0],
            to[1] - from[1]
        ]
    }
}


export type HexGridCoord = {
    x: number,
    y: number
}


export const HexGridCoord = {
    // Find the ends of the edges of a hexagon
    *ends(gc: HexGridCoord): Generator<EndRef, void, void> {

        let vertex = gc.y % 2 == 0 ?
            new TileRef(gc.x, Math.floor(gc.y / 2)).vertex(VertexClass.DEF) :
            new TileRef(gc.x+1, Math.floor(gc.y / 2)).vertex(VertexClass.ABC);

        // Get the top end of the vertical below the vertex
        const below = VertexRef.ends_cw(vertex)[0];

        let current = EndRef.lr(below)[1];

        for (let i = 0; i < 6; i++) {
            yield current
            current = EndRef.other(current);
            yield current
            current = EndRef.lr(current)[0]
        }
    }
}

// Can I translate a VertexRef and/or convert to/from a 3-part coord?
// Moving it to the left/right is easy - just add to the tile x
// For the next row down, need to change from a DEF to an ABC vertex
// and subtract 1 from y.

// It's always either an ABC or a DEF hexagon

// Can we go from a vertex to a physical offset
// Easy. The vertex class just has some implicit offset.
// For a given edge, can get a vector from one end to the other.


// Can we map to the 3-part coord system
// Then we can do rotation and translation in those directions


/* 
const v = TileRef.from(1, 0).vertex(VertexClass.CDE);

console.log(v);

const ve = VertexRef.ends_cw(v)
console.log(ve[0]);
const er = ve[0];

console.log(EndRef.other(er))

const lr = EndRef.lr(EndRef.other(er));
console.log(lr);

const other = EdgeRef.vertex_pair(lr[0]);

console.log(other)

console.log([...HexGridCoord.ends({x: 0, y: -1})])

const module_x = er.tile.x / 2;
const module_y = er.tile.y; */
// That decides where the frame is addressed to

// Given an EndRef, we can access the 4 pixels in the appropriate
// module frame.
// Each (tilex%2, edge class) is mapped to some
// location in the 4 neopixel channels.